qrenderdoc/Windows/MainWindow.cpp

```diff
@@ -2836,3 +2846,26 @@ void MainWindow::showLaunchError(ReplayStatus status)
   }
   GUIInvoke::call(this, [this, title, message]() { RDDialog::warning(this, title, message); });
 }
+
+bool MainWindow::isCapturableAppRunningOnAndroid(QString packageName)
+{
+  if(!m_Ctx.Replay().CurrentRemote() || !m_Ctx.Replay().CurrentRemote()->IsADB())
+    return false;
+
+  rdcstr host = m_Ctx.Replay().CurrentRemote()->hostname;
+  uint32_t ident = RENDERDOC_EnumerateRemoteTargets(host.c_str(), 0);
+
+  if(ident == 0)
+    return false;
+
+  ITargetControl *conn = RENDERDOC_CreateTargetControl(host.c_str(), ident, "testConnection", false);
```

Comment:
Because you're passing `forceConnection == false` here this may return `NULL` in the even that the connection is busy and a panel is already open. That's good as we don't want to disconnect an existing connection, but it means we might have false negatives. If a user tries to launch a new package while a connection to the existing one is going we wouldn't detect that.

Since the `RENDERDOC_EnumerateRemoteTargets` returning a valid non-0 ident already means that something is listening on that port I think that's enough, and we don't need to check if we can successfully make a target control connection.

Reply:
`RENDERDOC_EnumerateRemoteTargets` always returns a valid non-0 ident as it will finds the adb forward port every time.

Reply:
Connection is created by adb even though there isn't any application running on the device, so we have to check if we can make a successful connection.
Here is an example when no app is running:
$ adb forward --list
98885135374b4e4837 tcp:40020 localabstract:renderdoc_39920
98885135374b4e4837 tcp:39020 localabstract:renderdoc_38920
98885135374b4e4837 tcp:39970 localabstract:renderdoc_39920
98885135374b4e4837 tcp:38970 localabstract:renderdoc_38920
$ telnet localhost 39970
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
Connection closed by foreign host.

Reply:
39970 -> 39920 is the remote server forwarded port, so that seems like the renderdoc remote server will be listening on the other end regardless of whether any injected application is running. That's not part of the target control enumeration.

If I launch the remote server but have no running application, I can connect on 39970 (to the remote server) but not on 38970, which gets a connection refused because nothing is listening. If I run `EnumerateRemoteTargets` in the python shell with no application running, it returns 0 as I'd expect.

Reply:
I have also check it with the python shell and it returned with 38970 while no app was running.
```
>> renderdoc.EnumerateRemoteTargets("adb:0:98885135374b4e4837", 0)
38970
```
Did you test it on Windows? Because I'm using Linux and I might get different result due to the different os.

Reply:
I think I see the problem, somehow android accepts the connection and then closes it almost immediately after. So it seems like it's maybe a race condition for whether or not the socket will seem to open or not.

I've added a huge hack (355b81f232bdb08fd4fdc2a66156fd9e16bd95d4) which will just wait 100ms on android after opening the socket and then check that it's still open, to hopefully detect this case. Android is awful.

Reply:
This solves the problem, thank you. I have modified my commit.